mock/py/mockbuild/mounts.py

```diff
@@ -143,10 +143,11 @@ def __init__(self, rootObj):
         self.managed_mounts = []  # mounts owned by mock
         self.user_mounts = []  # mounts injected by user
         self.essential_mounts = [
-            FileSystemMountPoint(filetype='proc',
-                                 device='mock_chroot_proc',
-                                 path=rootObj.make_chroot_path('/proc')),
-            # avoid problems with kernel restrictions on mounting a new sysfs
+            # avoid problems with kernel restrictions on mounting a new procfs, sysfs
+            BindMountPoint(srcpath='/proc',
+                           bindpath=rootObj.make_chroot_path('/proc'),
+                           recursive=True,
```

Comment:
I wonder whether we need the `recursive` here at all? Previously when we mounted `/proc` and `/sys`, it mounted just the top level and did not mounted anything within. E.g., `/proc/sys/fs/binfmt_misc`.
Why it worked in past and why we need to recursively traverse now?

Reply:
The idea is that a container system might bind-mount *over* parts of /proc and /sys to hide them from processes inside the container. To prevent the container from escaping from this, it's not allowed to mount a new copy of the filesystem, but it's *also* not allowed to bind mount the filesystem in a way that would reveal the hidden parts - thus the kernel bindmount of /proc will fail unless it's recursive. I can make the comment say more of that instead of just "kernel restrictions".